lib/guardian/plug/pipeline.ex

```diff
@@ -122,71 +122,69 @@ defmodule Guardian.Plug.Pipeline do
   handler so that it can be used within your pipeline and downstream.
   """
   defmacro __using__(opts \\ []) do
+    alias Guardian.Plug.Pipeline
+
     quote do
       use Plug.Builder
-      import Guardian.Plug.Pipeline
 
-      def init(opts) do
-        otp_app = Keyword.get(opts, :otp_app) ||
-          Keyword.get(unquote(opts), :otp_app)
-        if !otp_app do
-          raise ":otp_app not specified by #{to_string(__MODULE__)}"
-        end
+      import Pipeline
 
-        config = Application.get_env(otp_app, __MODULE__) || []
+      plug :put_modules
 
+      def init(options) do
         new_opts =
-          config
+          options
           |> Keyword.merge(unquote(opts))
-          |> Keyword.merge(opts)
+          |> app_config
 
-        module = Keyword.get(new_opts, :module)
-        error = Keyword.get(new_opts, :error_handler)
+        unless Keyword.has_key?(new_opts, :module),
+          do: raise_error(:module)
 
-        if !module, do: raise ":module not specified for #{to_string(__MODULE__)}"
-        if !error, do: raise ":error_handler not specified for #{to_string(__MODULE__)}"
+        unless Keyword.has_key?(new_opts, :error_handler),
+          do: raise_error(:error_handler)
 
         new_opts
       end
 
-      plug :put_modules
+      defp app_config(opts) do
+        case Keyword.get(opts, :otp_app) do
+          nil -> opts
+          otp_app -> Application.get_env(otp_app, __MODULE__) ++ opts
+        end
+      end
 
       defp put_modules(conn, opts) do
-        alias Guardian.Plug, as: GPlug
-        alias GPlug.{Pipeline}
-
         pipeline_opts = Keyword.take(opts, [:module, :error_handler, :key])
         Pipeline.call(conn, pipeline_opts)
       end
+
+      defp raise_error(key),
+        do: raise "Config `#{key}` is missing for #{__MODULE__}"
     end
   end
 
   def call(conn, opts) do
+    [module, error_handler, key] =
+      opts
+      |> Keyword.take([:module, :error_handler, :key])
+      |> Keyword.values
+
     conn
-    |> maybe_set_item(:guardian_module, Keyword.get(opts, :module))
-    |> maybe_set_item(
-      :guardian_error_handler,
-      Keyword.get(opts, :error_handler)
-    )
-    |> maybe_set_item(:guardian_key, Keyword.get(opts, :key))
+    |> put_module(module)
```

Comment:
this changes the behaviour significantly. If those things aren't set they will override the values to nil rather than leaving them alone.